# Random Forest (Model2)
    - Final Dataframe: happy_postg_final_m2

In [41]:
# Retrieve variables from other jupyter notebook
%store -r happy_postg
%store -r X_train_scaled
%store -r X_test_scaled
%store -r y_train
%store -r y_test
%store -r X_headings
%store -r X_scaler


In [43]:
X_test_scaled

array([[ 0.        ,  0.16086818,  0.43462228,  0.24209131,  0.58697999,
         0.73436565,  0.71028917],
       [ 0.        ,  0.61857142,  0.80195531,  0.69799712,  0.57535285,
         0.22515193,  0.88857528],
       [ 1.        ,  0.24277664,  0.49476531,  0.13697622,  0.52391498,
         0.27117352,  0.86931155],
       [ 0.        ,  0.81168946,  0.79194107,  0.70800752,  0.88813881,
         0.47849813,  0.76703279],
       [ 1.        ,  0.51768965,  0.67813234,  0.30381919,  0.56228081,
         0.13314376,  0.89264877],
       [ 1.        ,  0.31009831,  0.76006229,  0.56227036,  0.75934164,
         0.24674348,  0.5495916 ],
       [ 1.        ,  0.26101363,  0.32895469,  0.24195257,  0.52391498,
         0.43119033,  0.79813302],
       [ 0.        ,  0.48214569,  0.81937988,  0.63910589,  0.39535166,
         0.10860315,  0.88102497],
       [ 1.        ,  0.6467067 ,  0.86345008,  0.78054795,  0.79770747,
         0.34446366,  0.86697783],
       [ 0.        ,  0.7929

In [2]:
# Test to ensure variables are retrived
y_train

3      8
127    5
249    6
1      8
245    6
      ..
70     6
132    5
289    5
109    5
176    7
Name: score, Length: 226, dtype: int64

In [3]:
# Dependencies
from sklearn.ensemble import RandomForestClassifier
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

In [4]:
# Create a random forest classifier
rf = RandomForestClassifier(n_estimators=770)
rf = rf.fit(X_train_scaled, y_train)
rf.score(X_test_scaled, y_test)

0.7631578947368421

# Save the model

In [5]:
# Save the svc model 
import joblib
filename = 'Models/model2.sav'

# Print file name
joblib.dump(rf, filename)

['Models/model2.sav']

# Additional checks on feature importances (X variables)

In [6]:
# Random Forests in sklearn will automatically calculate feature importance
importances = rf.feature_importances_
importances

array([0.01241298, 0.19585595, 0.19987538, 0.178343  , 0.15339349,
       0.11252278, 0.14759643])

In [7]:
X_headings

Index(['year', 'logged_GDP_per_capita', 'support', 'life_exp', 'freedom',
       'generosity', 'corruption'],
      dtype='object')

In [8]:
sorted(zip(rf.feature_importances_,X_headings), reverse=True)

[(0.19987537671784644, 'support'),
 (0.19585595270283346, 'logged_GDP_per_capita'),
 (0.17834300178411563, 'life_exp'),
 (0.1533934899366418, 'freedom'),
 (0.14759642512932394, 'corruption'),
 (0.11252277651550291, 'generosity'),
 (0.01241297721373579, 'year')]

In [9]:
# Test saved model loads smoothly
loaded_model = joblib.load('Models/model2.sav')

# Predict based on model2
predictions = loaded_model.predict(X_test_scaled)

# Put y_actual, y_predicted into dataframe
y_test_all_m2 = pd.DataFrame({"y_actual": y_test, "y_predicted": predictions})
y_test_all_m2

y_actual  y_predicted
112         5            5
63          6            6
269         5            5
39          7            7
276         5            5
..        ...          ...
214         6            7
187         7            6
296         4            4
46          7            6
114         5            5

[76 rows x 2 columns]

In [10]:
y_test_all_m2 = y_test_all_m2.reset_index()

In [11]:
# Merge 'happy_postg' on index with 'y_test_all_m2' on column name 'index'. (join=outer)
happy_postg_final_m2 = pd.merge(happy_postg,y_test_all_m2, how="outer", left_index=True, right_on="index")
happy_postg_final_m2 = happy_postg_final_m2.dropna()
happy_postg_final_m2 = happy_postg_final_m2.sort_index()
happy_postg_final_m2

year      country  logged_GDP_per_capita   support   life_exp   freedom  \
0.0   2020       Gambia               7.321815  0.693169  55.012016  0.733163   
1.0   2020       Serbia               9.680981  0.881476  68.210205  0.726496   
2.0   2021         Mali               7.744000  0.724000  51.969000  0.697000   
3.0   2020      Bahrain              10.676380  0.876342  68.500000  0.905856   
4.0   2021      Namibia               9.161000  0.818000  56.799000  0.719000   
...    ...          ...                    ...       ...        ...       ...   
71.0  2021  South Korea              10.651000  0.799000  73.900000  0.672000   
72.0  2021       Brazil               9.577000  0.882000  66.601000  0.804000   
73.0  2021       Malawi               6.958000  0.537000  57.948000  0.780000   
74.0  2020      Romania              10.107584  0.825162  67.207237  0.842823   
75.0  2020      Nigeria               8.576625  0.737217  49.861908  0.759578   

      generosity  corruption  score  index  y_actual  y_predicted  
0.0     0.343199    0.690718    5.0    112       5.0          5.0  
1.0    -0.073676    0.843509    6.0     63       6.0          6.0  
2.0    -0.036000    0.827000    5.0    269       5.0          5.0  
3.0     0.133729    0.739347    7.0     39       7.0          7.0  
4.0    -0.149000    0.847000    5.0    276       5.0          5.0  
...          ...         ...    ...    ...       ...          ...  
71.0   -0.083000    0.727000    6.0    214       6.0          7.0  
72.0   -0.071000    0.756000    7.0    187       7.0          6.0  
73.0    0.038000    0.729000    4.0    296       4.0          4.0  
74.0   -0.197815    0.934300    7.0     46       7.0          6.0  
75.0    0.033746    0.861874    5.0    114       5.0          5.0  

[76 rows x 12 columns]

In [13]:
# Rename column 'index' to 'id'
happy_postg_final_m2 = happy_postg_final_m2.rename({'index':'id','logged_GDP_per_capita':'logged_gdp_per_capita'}, axis='columns')

# Reorder column names
happy_postg_final_m2 = happy_postg_final_m2[['id','year','country','logged_gdp_per_capita','support','life_exp','freedom','generosity','corruption', 'y_actual','y_predicted']]
happy_postg_final_m2

id  year      country  logged_gdp_per_capita   support   life_exp  \
0.0   112  2020       Gambia               7.321815  0.693169  55.012016   
1.0    63  2020       Serbia               9.680981  0.881476  68.210205   
2.0   269  2021         Mali               7.744000  0.724000  51.969000   
3.0    39  2020      Bahrain              10.676380  0.876342  68.500000   
4.0   276  2021      Namibia               9.161000  0.818000  56.799000   
...   ...   ...          ...                    ...       ...        ...   
71.0  214  2021  South Korea              10.651000  0.799000  73.900000   
72.0  187  2021       Brazil               9.577000  0.882000  66.601000   
73.0  296  2021       Malawi               6.958000  0.537000  57.948000   
74.0   46  2020      Romania              10.107584  0.825162  67.207237   
75.0  114  2020      Nigeria               8.576625  0.737217  49.861908   

       freedom  generosity  corruption  y_actual  y_predicted  
0.0   0.733163    0.343199    0.690718       5.0          5.0  
1.0   0.726496   -0.073676    0.843509       6.0          6.0  
2.0   0.697000   -0.036000    0.827000       5.0          5.0  
3.0   0.905856    0.133729    0.739347       7.0          7.0  
4.0   0.719000   -0.149000    0.847000       5.0          5.0  
...        ...         ...         ...       ...          ...  
71.0  0.672000   -0.083000    0.727000       6.0          7.0  
72.0  0.804000   -0.071000    0.756000       7.0          6.0  
73.0  0.780000    0.038000    0.729000       4.0          4.0  
74.0  0.842823   -0.197815    0.934300       7.0          6.0  
75.0  0.759578    0.033746    0.861874       5.0          5.0  

[76 rows x 11 columns]

# LOAD to PostGRES

### Ensure pgAdmin is set up and create a table in database in
    - Develop environnment (PostGRES Database)
    - Deploy environment (Heroku PostGRES Database)



## Development Environment

In [ ]:
# Create a database called 'the_flow_db' in postgres database. Add the tables (table headings defined) into postgres database as per schema.sql (via a query in pgAdmin). 

In [14]:
from sqlalchemy import create_engine

# Development Environment - Connect to database 
rds_connection_string = "postgres:postgres@localhost:5432/the_flow_db"
engine = create_engine(f'postgresql://{rds_connection_string}')


In [15]:
# Add table headings via pgAdmin as per schema

# Run code to ensure connection is established and table has been created on postgres.
engine.table_names()

['happy_table_m2']

In [17]:
# Load dataframe (variable 'happy_postg_final_m2') into postgres database 'the_flow_db', in the table 'happy_table'
happy_postg_final_m2.to_sql(name='happy_table_m2', con=engine, if_exists='append', index=False)

In [27]:
# Run code to check connection is established and data is reading out from postgres database
m2_table = pd.read_sql_table('happy_table_m2', engine) 
m2_table

id  year      country  logged_gdp_per_capita   support   life_exp  \
0   112  2020       Gambia               7.321815  0.693169  55.012016   
1    63  2020       Serbia               9.680981  0.881476  68.210205   
2   269  2021         Mali               7.744000  0.724000  51.969000   
3    39  2020      Bahrain              10.676380  0.876342  68.500000   
4   276  2021      Namibia               9.161000  0.818000  56.799000   
..  ...   ...          ...                    ...       ...        ...   
71  214  2021  South Korea              10.651000  0.799000  73.900000   
72  187  2021       Brazil               9.577000  0.882000  66.601000   
73  296  2021       Malawi               6.958000  0.537000  57.948000   
74   46  2020      Romania              10.107584  0.825162  67.207237   
75  114  2020      Nigeria               8.576625  0.737217  49.861908   

     freedom  generosity  corruption  y_actual  y_predicted  
0   0.733163    0.343199    0.690718       5.0          5.0  
1   0.726496   -0.073676    0.843509       6.0          6.0  
2   0.697000   -0.036000    0.827000       5.0          5.0  
3   0.905856    0.133729    0.739347       7.0          7.0  
4   0.719000   -0.149000    0.847000       5.0          5.0  
..       ...         ...         ...       ...          ...  
71  0.672000   -0.083000    0.727000       6.0          7.0  
72  0.804000   -0.071000    0.756000       7.0          6.0  
73  0.780000    0.038000    0.729000       4.0          4.0  
74  0.842823   -0.197815    0.934300       7.0          6.0  
75  0.759578    0.033746    0.861874       5.0          5.0  

[76 rows x 11 columns]

## Deployment Environment

In [37]:
from sqlalchemy import create_engine

# Development Environment - Connect to database 
rds_connection_string = "zrhfplbtdfzlsb:7a52354918a924ca23e8e6f09d7e221ec33c71c911e49ac91fef71daac2db1dc@ec2-54-74-14-109.eu-west-1.compute.amazonaws.com:5432/dasrk3uqfvvv2e"
engine = create_engine(f'postgresql://{rds_connection_string}')

In [38]:
# Add table headings via pgAdmin as per schema

# Run code to ensure connection is established and table has been created on postgres.
engine.table_names()

['happy_table_m2']

In [35]:
# Load dataframe (variable 'happy_postg_final_m2') into postgres database 'the_flow_db', in the table 'happy_table'
happy_postg_final_m2.to_sql(name='happy_table_m2', con=engine, if_exists='append', index=False)

In [36]:
# Run code to check connection is established and data is reading out from postgres database
m2_table = pd.read_sql_table('happy_table_m2', engine) 
m2_table

id  year      country  logged_gdp_per_capita   support   life_exp  \
0   112  2020       Gambia               7.321815  0.693169  55.012016   
1    63  2020       Serbia               9.680981  0.881476  68.210205   
2   269  2021         Mali               7.744000  0.724000  51.969000   
3    39  2020      Bahrain              10.676380  0.876342  68.500000   
4   276  2021      Namibia               9.161000  0.818000  56.799000   
..  ...   ...          ...                    ...       ...        ...   
71  214  2021  South Korea              10.651000  0.799000  73.900000   
72  187  2021       Brazil               9.577000  0.882000  66.601000   
73  296  2021       Malawi               6.958000  0.537000  57.948000   
74   46  2020      Romania              10.107584  0.825162  67.207237   
75  114  2020      Nigeria               8.576625  0.737217  49.861908   

     freedom  generosity  corruption  y_actual  y_predicted  
0   0.733163    0.343199    0.690718       5.0          5.0  
1   0.726496   -0.073676    0.843509       6.0          6.0  
2   0.697000   -0.036000    0.827000       5.0          5.0  
3   0.905856    0.133729    0.739347       7.0          7.0  
4   0.719000   -0.149000    0.847000       5.0          5.0  
..       ...         ...         ...       ...          ...  
71  0.672000   -0.083000    0.727000       6.0          7.0  
72  0.804000   -0.071000    0.756000       7.0          6.0  
73  0.780000    0.038000    0.729000       4.0          4.0  
74  0.842823   -0.197815    0.934300       7.0          6.0  
75  0.759578    0.033746    0.861874       5.0          5.0  

[76 rows x 11 columns]

In [28]:
import json
m2_table_json = json.dumps(json.loads(m2_table.to_json(orient = "records")), indent=4)
m2_table_json = json.loads(m2_table_json)
m2_table_json

# m2_table_result = json.loads(test.to_json(orient = "records"), indent=4)
# m2_table_result = json.loads(m2_table_result)
# m2_table_result
    # return jsonify(m2_table_result)

[{'id': 112,
  'year': 2020,
  'country': 'Gambia',
  'logged_gdp_per_capita': 7.321814537,
  'support': 0.6931687,
  'life_exp': 55.0120163,
  'freedom': 0.733163178,
  'generosity': 0.343198717,
  'corruption': 0.690717816,
  'y_actual': 5.0,
  'y_predicted': 5.0},
 {'id': 63,
  'year': 2020,
  'country': 'Serbia',
  'logged_gdp_per_capita': 9.680980682,
  'support': 0.881475747,
  'life_exp': 68.21020508,
  'freedom': 0.726495862,
  'generosity': -0.073676221,
  'corruption': 0.843509018,
  'y_actual': 6.0,
  'y_predicted': 6.0},
 {'id': 269,
  'year': 2021,
  'country': 'Mali',
  'logged_gdp_per_capita': 7.744,
  'support': 0.724,
  'life_exp': 51.969,
  'freedom': 0.697,
  'generosity': -0.036,
  'corruption': 0.827,
  'y_actual': 5.0,
  'y_predicted': 5.0},
 {'id': 39,
  'year': 2020,
  'country': 'Bahrain',
  'logged_gdp_per_capita': 10.67638016,
  'support': 0.876342118,
  'life_exp': 68.5,
  'freedom': 0.905855775,
  'generosity': 0.13372919,
  'corruption': 0.7393471,
  'y_ac

## Deployment Environment

In [ ]:
# Create Heroku app (with Heroku Postgres). In pgAdmin, add this Heroku Postgres server. 
# Create a database called 'the_flow_db' in the Heroku postgres database. Add the tables (table headings defined) into postgres database as per schema.sql (via a query in pgAdmin). 

In [ ]:
# Deployment Environment
# rds_connection_string = "postgres:postgres@localhost:5432/events_db" NEED TO EDIT THIS 
engine = create_engine(f'postgresql://{rds_connection_string}')




# Test for 1 set of X values


# Test for 1 X_variables entry to get 1 y prediction

# Index(['year', 'logged_GDP_per_capita', 'support', 'life_exp', 'freedom',
#        'generosity', 'corruption'],
#       dtype='object')

yr = input
GDP_per_cap = input
logged_GDP_per_cap = natural log of GDP_per_cap
support = input
life_exp = input
freedom = input
gen = input
corrupt = input

X_inp = [yr, logged_GDP_per_cap, support, life_exp, freedom, gen, corrupt]





X_input = [[ 0.        ,  0.16086818,  0.43462228,  0.24209131,  0.58697999,
         0.73436565,  0.71028917]]

predictions = loaded_model.predict(X_input)
predictions